File: leetcode-learn/sorting/maximumGap.go
// You must write an algorithm
// that runs in linear time and uses linear extra space.

func maximumGap(nums []int) int {
	if len(nums) == 1 {
		return 0
	}

	// Counting Sort: O(N)
	maxValue := slices.Max(nums)
	shift := slices.Min(nums)
	count := make([]int, maxValue-shift+1)
	for _, n := range nums {
		count[n-shift] += 1
	}
	// find max diff
	maxDiff := 0
	prev := shift
	for n, c := range count {
		if c == 0 {
			continue
		}
		diff := n - prev
		if diff > maxDiff {
			maxDiff = diff
		}
		prev = n
	}
	return maxDiff
} // -> fatal error: runtime: out of memory

File: leetcode-learn/sorting/maximumGap.py
class Solution:
    def maximumGap(self, nums: List[int]) -> int:
        if len(nums) == 0: return 0
        nums.sort()
        max_diff = 0
        
        for i in range(len(nums) - 1):
            diff = nums[i + 1] - nums[i]
            if diff > max_diff:
                max_diff = diff
        return max_diff
